### PR TITLE
[Improve][CDC] Close idle subtasks gorup(reader/writer) in increment phase

### LINKED ISSUE #6542: [Bug] [mysql cdc] MySQL process cannot be released
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.

### What happened

作业创建后经过多次停作业，加表，恢复作业后业务库的连接进程越来越多，释放不了，停掉作业也不管用，重启seatunnel可释放。
image

### SeaTunnel Version

seatunnel:2.3.4

### SeaTunnel Config

```conf
env {
  execution.parallelism = 2
  job.mode = "STREAMING"
  checkpoint.interval = 3000
}
source {
    MySQL-CDC {
        parallelism = 1
        username = "xxx"
        password = "xxxx"
        base-url = "jdbc:mysql://xxxx"
        table-names = ["xfs_product_base.area","xfs_product_base.bi_test","xfs_product_base.check_standard_detail","xfs_product_base.check_standard_detail_audit","xfs_product_base.check_standard_sku_rel","xfs_product_base.check_standard_update_log","xfs_product_base.data_batch","xfs_product_base.product_accept_item_category_rel","xfs_product_base.product_add_state","xfs_product_base.product_alias","xfs_product_base.product_alias_name","xfs_product_base.product_area_salemin","xfs_product_base.product_brand","xfs_product_base.product_brand_audit","xfs_product_base.product_brand_series","xfs_product_base.product_category_brand_rel","xfs_product_base.product_category_platform","xfs_product_base.product_cate_temp_rel","xfs_product_base.product_classification","xfs_product_base.product_classification_cate_rel","xfs_product_base.product_collect","xfs_product_base.product_collect_potential_saleprice","xfs_product_base.product_collect_potential_supprice","xfs_product_base.product_collect_standard","xfs_product_base.product_dev_plan","xfs_product_base.product_enable_sku_record","xfs_product_base.product_enable_sku_record_detail","xfs_product_base.product_exclusive_main","xfs_product_base.product_exclusive_sku","xfs_product_base.product_keyword","xfs_product_base.product_label_sku_rel","xfs_product_base.product_main","xfs_product_base.product_main_description","xfs_product_base.product_main_image","xfs_product_base.product_main_multi_image","xfs_product_base.product_management_attrinfo_audit","xfs_product_base.product_manager_marketing_sku_rel","xfs_product_base.product_not_spec_attrinfo","xfs_product_base.product_org","xfs_product_base.product_potential","xfs_product_base.product_potential_change_log","xfs_product_base.product_potential_record","xfs_product_base.product_potential_record_detail","xfs_product_base.product_potential_standard","xfs_product_base.product_potential_standard_detail","xfs_product_base.product_prepare_commit_record","xfs_product_base.product_prepare_commit_record_main","xfs_product_base.product_qualification","xfs_product_base.product_qualificationtypesku_qua_rel","xfs_product_base.product_qualificationtype_sku_rel","xfs_product_base.product_qualification_attachment","xfs_product_base.product_restrict_show_plat","xfs_product_base.product_scene_category","xfs_product_base.product_scene_category_spu","xfs_product_base.product_sku","xfs_product_base.product_sku_audit","xfs_product_base.product_sku_image","xfs_product_base.product_sku_package_detail","xfs_product_base.product_sku_package_main","xfs_product_base.product_supply_price","xfs_product_base.product_supply_province_relate","xfs_product_base.product_supply_salearea_product","xfs_product_base.product_supply_salearea_sku","xfs_product_base.product_supply_salecity_product","xfs_product_base.product_supply_sku_rel","xfs_product_base.product_tax_category","xfs_product_base.product_template","xfs_product_base.product_template_attr","xfs_product_base.product_temporary_add_attrinfo","xfs_product_base.product_temporary_add_attrinfo_audit","xfs_product_base.product_twins_sku","xfs_product_base.product_unit","xfs_product_base.product_unit_sku_rel","xfs_product_base.product_update_commit_record","xfs_product_base.product_update_commit_record_limit_detail","xfs_product_base.product_updown_sku_record","xfs_product_base.product_updown_sku_record_detail","xfs_product_base.purchase_price_rule","xfs_product_base.raw_material_todo_list","xfs_product_base.temp_project_sku","xfs_product_base.temp_sku_spec"]
        result_table_name = "阿里云_xfs_product_base"
        startup.mode = "latest"
    }
}
sink {
    StarRocks {
        base-url = "jdbc:mysql://xxxx"
        username = "xxx"
        password = "xxx"
        database = "xfs_product_base"
        source_table_name = ["阿里云_xfs_product_base"]
        nodeUrls = ["xxxx"]
        enable_upsert_delete = true
    }
}
```

### Running Command

```shell
/data/seatunnel-2.3.4/bin/seatunnel.sh --async --config /data/seatunnel-2.3.4/job/xxx.conf
/data/seatunnel-2.3.4/bin/seatunnel.sh -s 822317531900411905 
/data/seatunnel-2.3.4/bin/seatunnel.sh -r 822317531900411905 --async --config /data/seatunnel-2.3.4/job/xxx.conf
/data/seatunnel-2.3.4/bin/seatunnel.sh -s 822317531900411905 
/data/seatunnel-2.3.4/bin/seatunnel.sh -r 822317531900411905 --async --config /data/seatunnel-2.3.4/job/xxx.conf
```

### Error Exception

```log
no errors
```

### Zeta or Flink or Spark Version

_No response_

### Java or Scala Version

_No response_

### Screenshots

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by hailin0**
### Purpose of this pull request

[CDC] Close idle subtasks gorup(reader/writer) in increment phase

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
added
https://github.com/apache/seatunnel/pull/6526/files#diff-198d9ea0204d7713e1623ecd32a24c516808b469f8f792211e3f71b0a101b385R421

https://github.com/apache/seatunnel/pull/6526/files#diff-5f0bf5f7076dac7f98979b1221bd1be4e3c24926beb6322d987a06a9a8b6b2e3R229

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-connectors-v2/connector-cdc/connector-cdc-mysql/src/main/java/org/apache/seatunnel/connectors/seatunnel/cdc/mysql/source/MysqlPooledDataSourceFactory.java [hunk 1]
```diff
@@ -1,36 +0,0 @@
-/*
- * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-cdc/connector-cdc-mysql/src/main/java/org/apache/seatunnel/connectors/seatunnel/cdc/mysql/source/MysqlPooledDataSourceFactory.java:2]: link
https://github.com/apache/seatunnel/pull/6556

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/SeaTunnelTask.java [hunk 2]
```diff
@@ -339,7 +339,8 @@ public void ack(Barrier barrier) {
         Integer ackSize =
                 cycleAcks.compute(barrier.getId(), (id, count) -> count == null ? 1 : ++count);
         if (ackSize == allCycles.size()) {
-            if (barrier.prepareClose()) {
+            cycleAcks.remove(barrier.getId());
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/task/SeaTunnelTask.java:342]: cc @Hisoka-X

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-cdc-mysql-e2e/src/test/java/org/apache/seatunnel/connectors/seatunnel/cdc/mysql/MysqlCDCIT.java [hunk 3]
```diff
@@ -412,6 +418,13 @@ public void testMultiTableWithRestore(TestContainer container)
                                                                         MYSQL_DATABASE2,
                                                                         SOURCE_TABLE_2)))));
 
+        await().atMost(60000, TimeUnit.MILLISECONDS)
+                .pollInterval(1000, TimeUnit.MILLISECONDS)
+                .until(() -> getConnectionStatus("st_user_source").size() == 1);
+        await().atMost(60000, TimeUnit.MILLISECONDS)
+                .pollInterval(1000, TimeUnit.MILLISECONDS)
+                .until(() -> getConnectionStatus("st_user_sink").size() == 1);
+
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-cdc-mysql-e2e/src/test/java/org/apache/seatunnel/connectors/seatunnel/cdc/mysql/MysqlCDCIT.java:428]: check connection count

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-cdc-mysql-e2e/src/test/resources/mysqlcdc_to_mysql_with_multi_table_mode_one_table.conf [hunk 4]
```diff
@@ -20,19 +20,22 @@
 
 env {
   # You can set engine configuration here
-  parallelism = 1
+  parallelism = 3
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-cdc-mysql-e2e/src/test/resources/mysqlcdc_to_mysql_with_multi_table_mode_one_table.conf:23]: init parallelism = 3

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-cdc-mysql-e2e/src/test/resources/mysqlcdc_to_mysql_with_multi_table_mode_one_table.conf [hunk 5]
```diff
@@ -20,19 +20,22 @@
...
-    server-id = 5652
-    username = "mysqluser"
+    server-id = 5652-5660
+    username = "st_user_source"
     password = "mysqlpw"
     table-names = ["mysql_cdc.mysql_cdc_e2e_source_table"]
     base-url = "jdbc:mysql://mysql_cdc_e2e:3306/mysql_cdc"
+
+    snapshot.split.size = 1
+    snapshot.fetch.size = 1
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-cdc-mysql-e2e/src/test/resources/mysqlcdc_to_mysql_with_multi_table_mode_one_table.conf:38]: init parallelism=3 and snapshot.split.size = 1

Make sure that all 3 readers can open jdbc connections to read data

File: seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/base/source/enumerator/HybridSplitAssigner.java [hunk 6]
```diff
@@ -114,7 +114,8 @@ public Optional<SourceSplitBase> getNext() {
 
     @Override
     public boolean waitingForCompletedSplits() {
-        return snapshotSplitAssigner.waitingForCompletedSplits();
+        return snapshotSplitAssigner.waitingForCompletedSplits()
+                || incrementalSplitAssigner.waitingForAssignedSplits();
```
Comment (dailai) [inline seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/base/source/enumerator/HybridSplitAssigner.java:118]: Should we keep the called method named incrementalSplitAssigner.waitingForCompletedSplits() ?

File: seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/base/source/enumerator/IncrementalSplitAssigner.java [hunk 7]
```diff
@@ -278,4 +278,8 @@ public boolean completedSnapshotPhase(List<TableId> tableIds) {
         return context.getAssignedSnapshotSplit().isEmpty()
                 && context.getSplitCompletedOffsets().isEmpty();
     }
+
+    public boolean waitingForAssignedSplits() {
```
Comment (dailai) [inline seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/base/source/enumerator/IncrementalSplitAssigner.java:282]: Keep the logical in waitingForCompletedSplits?

File: seatunnel-connectors-v2/connector-cdc/connector-cdc-mysql/src/main/java/org/apache/seatunnel/connectors/seatunnel/cdc/mysql/source/reader/fetch/binlog/MySqlBinlogFetchTask.java [hunk 8]
```diff
@@ -72,6 +76,22 @@ public void execute(FetchTask.Context context) throws Exception {
         BinlogSplitChangeEventSourceContext changeEventSourceContext =
                 new BinlogSplitChangeEventSourceContext();
 
+        sourceFetchContext
+                .getBinaryLogClient()
+                .registerLifecycleListener(
+                        new BinaryLogClient.AbstractLifecycleListener() {
+                            @Override
+                            public void onConnect(BinaryLogClient client) {
```
Comment (dailai) [inline seatunnel-connectors-v2/connector-cdc/connector-cdc-mysql/src/main/java/org/apache/seatunnel/connectors/seatunnel/cdc/mysql/source/reader/fetch/binlog/MySqlBinlogFetchTask.java:84]: Should we need to rewrite the method named onDisconnect?

Comment (dailai) [reply seatunnel-connectors-v2/connector-cdc/connector-cdc-mysql/src/main/java/org/apache/seatunnel/connectors/seatunnel/cdc/mysql/source/reader/fetch/binlog/MySqlBinlogFetchTask.java:84]: When we stop the task, the method named org.apache.seatunnel.connectors.seatunnel.cdc.mysql.source.reader.fetch.MySqlSourceFetchTaskContext#close will be called.

File: seatunnel-connectors-v2/connector-cdc/connector-cdc-postgres/src/main/java/org/apache/seatunnel/connectors/seatunnel/cdc/postgres/source/reader/PostgresSourceFetchTaskContext.java [hunk 9]
```diff
@@ -180,28 +180,29 @@ public void configure(SourceSplitBase sourceSplitBase) {
...
+                    createReplicationConnection(
+                            doSnapshot, connectorConfig.maxRetries(), connectorConfig.retryDelay());
+                }
                 // we need to create the slot before we start streaming if it doesn't exist
                 // otherwise we can't stream back changes happening while the snapshot is taking
                 // place
                 if (slotInfo == null) {
                     try {
-                        slotCreatedInfo =
+                        SlotCreationResult slotCreatedInfo =
```
Comment (dailai) [inline seatunnel-connectors-v2/connector-cdc/connector-cdc-postgres/src/main/java/org/apache/seatunnel/connectors/seatunnel/cdc/postgres/source/reader/PostgresSourceFetchTaskContext.java]: Please remove useless variable definitions.

File: seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/source/reader/SourceReaderBase.java [hunk 10]
```diff
@@ -146,9 +147,15 @@ public void handleSourceEvent(SourceEvent sourceEvent) {
         log.info("Received unhandled source event: {}", sourceEvent);
     }
 
+    protected boolean isNoMoreElement() {
```
Comment (Carl-Zhou-CN) [inline seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/source/reader/SourceReaderBase.java:150]: Should we limit the scope of this feature to within the CDC?

File: seatunnel-connectors-v2/connector-cdc/connector-cdc-postgres/src/main/java/org/apache/seatunnel/connectors/seatunnel/cdc/postgres/source/reader/PostgresSourceFetchTaskContext.java (hunk 9 above)
Comment (hailin0) [reply seatunnel-connectors-v2/connector-cdc/connector-cdc-postgres/src/main/java/org/apache/seatunnel/connectors/seatunnel/cdc/postgres/source/reader/PostgresSourceFetchTaskContext.java]: fix by this pr
https://github.com/apache/seatunnel/pull/6785/files#diff-68a4e18adfe57dda2d37a3fc3092effa5c1a326076a16bb88131e82a62dba38fL193

File: seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/base/source/enumerator/IncrementalSplitAssigner.java (hunk 7 above)
Comment (hailin0) [reply seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/base/source/enumerator/IncrementalSplitAssigner.java:282]: Stream log reading never ends, so not use waitingForCompletedSplits

File: seatunnel-connectors-v2/connector-cdc/connector-cdc-mysql/src/main/java/org/apache/seatunnel/connectors/seatunnel/cdc/mysql/source/reader/fetch/binlog/MySqlBinlogFetchTask.java (hunk 8 above)
Comment (hailin0) [reply seatunnel-connectors-v2/connector-cdc/connector-cdc-mysql/src/main/java/org/apache/seatunnel/connectors/seatunnel/cdc/mysql/source/reader/fetch/binlog/MySqlBinlogFetchTask.java:84]: This connection allows repeated closing

Comment (hailin0) [reply seatunnel-connectors-v2/connector-cdc/connector-cdc-mysql/src/main/java/org/apache/seatunnel/connectors/seatunnel/cdc/mysql/source/reader/fetch/binlog/MySqlBinlogFetchTask.java:84]: No need to override onDisconnect method

File: seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/base/source/enumerator/HybridSplitAssigner.java (hunk 6 above)
Comment (hailin0) [reply seatunnel-connectors-v2/connector-cdc/connector-cdc-base/src/main/java/org/apache/seatunnel/connectors/cdc/base/source/enumerator/HybridSplitAssigner.java:118]: will use it in the future

File: seatunnel-e2e/seatunnel-e2e-common/src/test/resources/log4j2.properties [hunk 11]
```diff
@@ -16,7 +16,7 @@
 # limitations under the License.
 ################################################################################
 
-rootLogger.level = INFO
+rootLogger.level = WARN
```
Comment (dailai) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/resources/log4j2.properties]: Why do not keep the log level to INFO?

File: seatunnel-e2e/seatunnel-engine-e2e/connector-seatunnel-e2e-base/src/test/resources/log4j2.properties [hunk 12]
```diff
@@ -16,7 +16,7 @@
 # limitations under the License.
 ################################################################################
 
-rootLogger.level = INFO
+rootLogger.level = WARN
```
Comment (dailai) [inline seatunnel-e2e/seatunnel-engine-e2e/connector-seatunnel-e2e-base/src/test/resources/log4j2.properties]: As above

**Review by dailai — APPROVED** [review]
LGTM

**Comment by dailai** [comment]
Good job.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-http-e2e/src/test/java/org/apache/seatunnel/e2e/connector/http/HttpIT.java [hunk 13]
```diff
@@ -75,6 +75,7 @@ public void startUp() {
                         .withEnv(
                                 "MOCKSERVER_INITIALIZATION_JSON_PATH",
                                 TMP_DIR + getMockServerConfig())
+                        .withEnv("MOCKSERVER_LOG_LEVEL", "WARN")
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-http-e2e/src/test/java/org/apache/seatunnel/e2e/connector/http/HttpIT.java:78]: disable info logs

File: seatunnel-connectors-v2/connector-http/connector-http-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/http/source/HttpSourceReader.java [hunk 14]
```diff
@@ -130,7 +130,7 @@ public void pollAndCollectData(Collector<SeaTunnelRow> output) throws Exception
                     collect(output, content);
                 }
             }
-            log.info(
+            log.debug(
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-http/connector-http-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/http/source/HttpSourceReader.java:133]: change to debug

File: seatunnel-e2e/seatunnel-e2e-common/src/test/resources/log4j2.properties (hunk 11 above)
Comment (hailin0) [reply seatunnel-e2e/seatunnel-e2e-common/src/test/resources/log4j2.properties]: fix e2e fail by logs

https://github.com/apache/seatunnel/pull/6526/files#diff-37ec4d574545a62fd6623d950e8ba947c72116663ada3424cd673ffbe2a733daR133
https://github.com/apache/seatunnel/pull/6526/files#diff-aa135f68a09a09ca7e083debab4cd5814bcf73d17361818532e68cd26ef263bcR78

**Review by Carl-Zhou-CN — APPROVED** [review]
+1

File: seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/source/reader/SourceReaderBase.java (hunk 10 above)
Comment (hailin0) [reply seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/source/reader/SourceReaderBase.java:150]: No, this is common logic of all connector

**Review by EricJoy2048 — APPROVED** [review]

**Pull request merged**